File: Data Structures & Algorithms/rotting-fruit/submission-3.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n = grid.size();
        int m = grid[0].size();

        int minute = 0;  // how many minutes have passed
        int fresh = 0;   // count of fresh fruits still left

        queue<pair<int,int>> q; // bfs queue storing positions of rotten fruits

        // 1) scan the grid
        //    - count how many fresh fruits we have
        //    - add all starting rotten fruits into the queue
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == 1) {
                    fresh++; // found a fresh fruit
                }
                if (grid[i][j] == 2) {
                    q.push({i, j}); // found a rotten fruit, start point for bfs
                }
            }
        }

        // directions for 4-neighbors: right, left, down, up
        vector<pair<int, int>> directions = {
            {0, 1}, {0, -1}, {1, 0}, {-1, 0}
        };

        // 2) bfs by "levels"
        // each level of bfs = 1 minute passing
        // stop if there are no fresh fruits left or no rotten fruits to spread from
        while (fresh > 0 && !q.empty()) {

            // length = how many rotten fruits exist at the start of this minute
            // we only process these this minute
            // newly rotten fruits get pushed and handled next minute
            int length = q.size();

            for (int i = 0; i < length; i++) {
                auto curr = q.front();
                q.pop();

                int r = curr.first;
                int c = curr.second;

                // try to rot the 4 neighbors
                for (const auto& dir : directions) {
                    int row = r + dir.first;
                    int col = c + dir.second;

                    // check bounds and check if neighbor is fresh
                    if (row >= 0 && row < n &&
                        col >= 0 && col < m &&
                        grid[row][col] == 1) {

                        // rot it
                        grid[row][col] = 2;

                        // it will spread rot in the next minute
                        q.push({row, col});

                        // one less fresh fruit
                        fresh--;
                    }
                }
            }

            // after processing one whole level, 1 minute has passed
            minute++;
        }

        // if we rotted everything, return minutes, otherwise impossible
        return (fresh == 0) ? minute : -1;
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-3.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []
        
        # Track boundaries
        top = 0
        bottom = len(matrix) - 1
        left = 0
        right = len(matrix[0]) - 1
        
        while top <= bottom and left <= right:
            # 1. Go RIGHT (top row, left to right)
            for col in range(left, right + 1):
                result.append(matrix[top][col])
            top += 1
            
            # 2. Go DOWN (right column, top to bottom)
            for row in range(top, bottom + 1):
                result.append(matrix[row][right])
            right -= 1
            
            # 3. Go LEFT (bottom row, right to left)
            # Check if there's still a row to process
            if top <= bottom:
                for col in range(right, left - 1, -1):
                    result.append(matrix[bottom][col])
                bottom -= 1
            
            # 4. Go UP (left column, bottom to top)
            # Check if there's still a column to process
            if left <= right:
                for row in range(bottom, top - 1, -1):
                    result.append(matrix[row][left])
                left += 1
        
        return result